Properly handling async model for initializing MediaCapture

LoadLibrary opened the pull request:
The documentation states that in apps that use C# or C++, the first use of the MediaCapture object to call InitializeAsync family methods should be on the STA thread. Calls from an MTA thread may result in undefined behavior.

It seems that, in UWP, calling MediaCapture::InitializeAsync from a MTA thread works, but calling MediaCapture::InitializeWithSettingsAsync from a MTA thread doesn’t work properly. This restriction seems to not repro in Win32.

This change transfers the MediaCapture initialization call to the STA thread (UI thread). Additionally, this change handles correctly the async pattern used for requesting camera and microphone permissions from the user.

LoadLibrary committed:
Properly handling async model for initializing MediaCapture

The documentation states that in apps that use C# or C++, the first use of the MediaCapture object to call InitializeAsync family methods should be on the STA thread. Calls from an MTA thread may result in undefined behavior.

It seems that, in UWP, calling MediaCapture::InitializeAsync from a MTA thread works, but calling MediaCapture::InitializeWithSettingsAsync from a MTA thread doesn’t work properly. This restriction seems to not repro in Win32.

This change transfers the MediaCapture initialization call to the STA thread (UI thread). Additionally, this change handles correctly the async pattern used for requesting camera and microphone permissions from the user.

LoadLibrary committed:
Working around the stride and gap issues with NV12 memory layout

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
+                                              &bitmap_capacity));
+ 
+-    pfn_incoming_frame_(bitmap_content, bitmap_capacity, frameInfo, 0);
++    // hack(auright): The following hack works around the limitations of 
```
"hack" doesn't pass PolyCheck, can't release with that.

LoadLibrary replied:
Replaced with "workaround"

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
+ 
+-    pfn_incoming_frame_(bitmap_content, bitmap_capacity, frameInfo, 0);
++    // hack(auright): The following hack works around the limitations of 
++    // libWebRTC memory representations for YUV planar images. Windows::Media::Capture allows
```
`libwerbtc` representation of a video frame is `webrtc::I420Buffer` which has separate Y, U, and V data pointers and strides, and therefore handles any kind of stride. Where is this limitation you are talking about? I've never seen anything like it.

jehumb-havok replied:
If the problem is `VideoCaptureImpl::IncomingFrame()` then it should probably be fixed/reimplemented with an overload that takes an actual strided frame. The implementation shows that it stores the frame content in a `webrtc::I420Buffer` so we should avoid making a copy just because of that one function signature. Since we're patching we can add an overload that takes an actual strided frame.

LoadLibrary replied:
As per your second comment, a more thoughtful read of the code would lead you to ``VideoCaptureImpl::IncomingFrame``.

An overload would be another solution, but we come down to the harsh reality of project schedules and what is good enough for now. The comment in the code makes clear that this inefficiency is purposely by design and temporarily.
Would you prefer to fix this code or to invest in porting Anaheim's code over to WinRTC?

BTW, WebRTC has other representation for video frames. Check api\video\video_frame_buffer.h.

jehumb-havok replied:
> An overload would be another solution, but we come down to the harsh reality of project schedules and what is good enough for now.

The amount of code to convert here is roughly equivalent to the amount of code that would have been necessary to write an overload in the first place.

> BTW, WebRTC has other representation for video frames. Check api\video\video_frame_buffer.h.

The base class `VideoFrameBuffer` has a mandatory `ToI420()` which returns a `I420BufferInterface` which supports separate Y, U, and V planes with arbitrary strides. This is the base interop format for all modules to communicate with each other; every format derived needs to support that.

fiban-havok replied:
@LoadLibrary 
> Would you prefer to fix this code or to invest in porting Anaheim's code over to WinRTC?

We'd prefer feature parity with WebRTC-UWP in the short term. If this code is introducing an extra copy compared to WebRTC-UWP, then it should be fixed.

LoadLibrary replied:
The problem wasn't related to `webrtc::I420Buffer`. I had to fix a bug in libyuv that wasn't taking YUV width as stride. Additionally, libyuv didn't have support for a gap between Y and UV planes (also fixed).

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
++                .Get(),
++            &init_settings));
++        THR(init_settings->put_StreamingCaptureMode(
++            StreamingCaptureMode::StreamingCaptureMode_AudioAndVideo));
```
This is a video capture module, it shouldn't require users to grant audio permission.

LoadLibrary replied:
This is a tricky one... I agree with you, but changing to StreamingCaptureMode_Video would require to change the Media Capture initialization code from the audio capture module for waiting for user granting permissions.

I could do that, but I'd prefer to use my time porting Anaheim's code over WinRTC. Arguably, the issue here is WebRTC built-in architecture that disassociates audio and video without having a common initialization step. Since most of the current use cases require audio and video, I don't think that the code as-is harmful.

Do you agree with this assessment?

jehumb-havok replied:
Currently on WebRTC UWP SDK the audio permission is required at startup because the audio device module (ADM; different from the capturer) initializes with the library (see `WebRtcVoiceEngine` init). The video permission on the other hand is only required if actually using video capture. This is the same on Android by the way; this is not specific to UWP, but rather to auth-based platforms in general. I don't know exactly how that works with the WinRT audio capturer you wrote but I am under the impression this won't change since this is an ADM limitation; so this audio permission is useless anyway, since it is requested much later, and you cannot reach that point in the code without crashing, at least with ADM1 (default). Feel free to leave for now, but I am pretty sure with or without won't make a difference, and if the current audio capture doesn't ask for permission then it is already susceptible to crashing the app, whether or not this change happens.

> Since most of the current use cases require audio and video

That is not always the case for Mixed Reality; several of our partners have apps with only one-way video, so video capture is not used on one peer. Some may have audio-only apps too. Although granted I do not know of any application using only video without audio.

LoadLibrary replied:
It has been fixed in latest update (it's using StreamingCaptureMode::StreamingCaptureMode_Video).

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
++            &init_settings5));
++        THR(init_settings5->put_MemoryPreference(
++            MediaCaptureMemoryPreference::MediaCaptureMemoryPreference_Cpu));
++        THR(init_settings->put_VideoDeviceId(device_id));
```
Why is the video device specified?
- This is undocumented (the API docs don't mention why one would pass a value here, they only seem to tell this is filled back)
- The old WebRTC UWP code doesn't pass anything and works fine

I have never been able to find any info on this, so no idea what are the consequences of passing a specific video device ID here.

LoadLibrary replied:
Video device ID is passed for allowing initializing multiple different cameras with different settings. Think about an app running on a Surface Laptop 3. The user might want to use the built-in camera that has access to the VRAM and/or it might want to use another webcam connected via USB that will transfer its frames to RAM.  

This code only uses public and documented APIs. This is a "put" field. API designer would set this property as read only if it would be the case. Feel free to suggest SiGMa team to update the documentation.

It seems that the old WebRTC UWP didn't check all the boxes.

jehumb-havok replied:
> Video device ID is passed for allowing initializing multiple different cameras with different settings.

I didn't find anything about this in the docs, where is that info coming from? Feel free to leave, I guess it doesn't hurt.

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
++
++        THR(media_capture_async.AsAgile(&async_action_media_capture_agile));
++
++      Cleanup:
```
No `goto` please.

LoadLibrary replied:
Not all uses of ``goto`` are evil.
As previously discussed, this code relies on the recommended "Jump on Fail" error handling pattern for COM published by Microsoft (https://docs.microsoft.com/en-us/windows/win32/learnwin32/error-handling-in-com#jump-on-fail).

jehumb-havok replied:
> As previously discussed, this code relies on the recommended "Jump on Fail" error handling pattern for COM published by Microsoft

One of four different suggested patterns, each with pros and cons. I appreciate this is your own preferred way, but `goto` is forbidden by Google/WebRTC and I am rarely myself find it justified.

https://docs.microsoft.com/en-us/windows/win32/learnwin32/error-handling-in-com#patterns-for-error-handling
> This section looks at some patterns for handling COM errors in a structured way. **Each pattern has advantages and disadvantages. To some extent, the choice is a matter of taste.**

LoadLibrary replied:
The "Cascade of ifs"pattern has been adopted in the latest update. No more macros or gotos.

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
+-using Microsoft::WRL::Wrappers::HStringReference;
+-using Microsoft::WRL::Wrappers::RoInitializeWrapper;
+-using std::vector;
++using ::ABI::Windows::ApplicationModel::Core::ICoreApplication;
```
I still don't understand why this code uses WinRT. Is there some value for you here? Neither us nor Google need WinRT. The original Chromium video capture doesn't use WinRT, which makes this code widely different and requires an entire rewriting instead of modifying the Google version to fit the WebRTC video track source interface.

jehumb-havok replied:
Beyond that, the code is widely different from `media/capture/video/win/video_capture_device_mf_win.cc` in Chromium. It doesn't even look like they're using the same API (can't find any reference to IMFMediaSample, media sink, etc.). It is virtually impossible to tell if it's doing the right thing and what are the differences in behavior with the Chromium video capturer. In short, this is a completely new and unrelated implementation.

fiban-havok replied:
My 2c: I agree with the points above, OTOH our first priority should be feature parity with WebRTC-UWP (as discussed), so it is "fine" if this unrelated implementation is a stopgap as long as it's quick and it works.

LoadLibrary replied:
This code was written before our agreement of having feature parity with Anaheim. This code was designed and implemented to replace the DircetShow, to work as turnkey solution for Win32 and UWP without changing WebRTC's built-in video module architecture. I'm ok replacing all this code after having the new code based on Anaheim ready to ship. Until then, this is the best effort for having video capabilities using modern Windows APIs without changing WebRTC built-in video capture module architecture.

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
++    hr = (expr);                                            \
++    if (FAILED(hr)) {                                       \
++      RTC_LOG(LS_ERROR) << "HR: " << hr;                    \
++      RTC_LOG(LS_ERROR) << __FUNCTION__ << ":" << __LINE__; \
```
Use `RTC_LOG_F` instead of rolling your own.

LoadLibrary replied:
Will do. Thanks.

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
+-    if (FAILED(hr)) \
+-      goto Cleanup; \
++#ifdef _DEBUG
++#define THR(expr)                                           \
```
From experience this kind of macro makes debugging extremely difficult. We banned them after a year on a large project, and reverted all uses, because it was causing so much pain to everyone.

fiban-havok replied:
IIRC the problem with `#define MACRO (...) if (FAILED(...)) return;` was that you couldn't break on the `return` i.e. on the failure. This has a `__debugbreak()` and it seems to be used mostly an assert rather than to validate input/handle expected error conditions, so maybe it's fine?

LoadLibrary replied:
The alternative would make the code harder to read. A considerable part of the OS components (which one could call as a large project...) and several third-party COM components rely on macros like this one.
Again, this is a consequence of using "Jump on Fail" error handling pattern for COM published by Microsoft (https://docs.microsoft.com/en-us/windows/win32/learnwin32/error-handling-in-com#jump-on-fail).

jehumb-havok replied:
> The alternative would make the code harder to read.

The page you linked only contains examples which do _not_ use such a macro in any of the 4 error handling patterns.

> A considerable part of the OS components (which one could call as a large project...) and several third-party COM components rely on macros like this one.

This code is old, I do not consider it a good reference on modern coding practices. Again, I have some experience debugging through the code you mention, especially via crash dumps, and being unable to set breakpoints makes things unnecessarily difficult. As Filippo pointed out the `__debugbreak()` here alleviates the issue though, so let's say this is OK.

LoadLibrary replied:
The change has been updated using "Cascade of ifs" instead of "Jump on fail". No more macros or gotos.

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
++      RTC_LOG(LS_ERROR) << "HR: " << hr;                    \
++      RTC_LOG(LS_ERROR) << __FUNCTION__ << ":" << __LINE__; \
++      __debugbreak();                                       \
++      goto Cleanup;                                         \
```
Google doesn't allow `goto`, so this code can never be pushed back to Goggle. And frankly I am vert reticent in taking any code with that construct.

LoadLibrary replied:
You'll find out that a guideline is just a guideline...
A quick search in Chromium code base found more than 3000+ uses of ``goto``. I didn't spend too much time looking at each use, but the beloved Protobuffer code has several uses of ``goto`` in its code and in the code that it generates. 

Including my favorite and so familiar:
``#define CHK_(x) if (PROTOBUF_PREDICT_FALSE(!(x))) goto failure``

Not all uses of ``goto`` are evil. I STRONGLY recommend a read of the chapter 16 of Steve McConnell's "Code Complete" book. Spoiler alert, page 351-352 will show you a very familiar code to this PR.

fiban-havok replied:
@LoadLibrary the use of `goto`s and `THR` macros has already caused friction when contributing code to Google, see [here](https://webrtc-review.googlesource.com/c/src/+/167021) for example.

I'd say if we plan to contribute to a codebase, let's be respectful and follow the guidelines of that codebase as well as possible. As you point out, guidelines are just guidelines and this includes the COM error reporting conventions - we should use our growth mindset and evaluate constructs on their merits in a specific situation rather than stick to them because they have been successfully used elsewhere. IMHO contributing back this code is a concern, and we should make it as frictionless as possible, and complex macros and `goto`s are not _really_ essential here, should they be avoided.

EDIT: ofc this assumes that we plan to contribute this code to WebRTC - if we plan to replace it completely and this is not the case then anything goes!

LoadLibrary replied:
@fibann You're right, the code has been updated in favor of "Cascade of ifs" instead of "Junp on fail". No more macros or gotos.

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
+   ComPtr<IMediaCapture5> media_capture5;
+   ComPtr<IMapView<HSTRING, MediaFrameSource*>> frame_sources;
+   ComPtr<IIterable<IKeyValuePair<HSTRING, MediaFrameSource*>*>> iterable;
+-  ComPtr<IIterator<IKeyValuePair<HSTRING, MediaFrameSource*>*>> frame_source_iterator;
```
I can't find where this is used. Neither here nor in the patch that created those files.

LoadLibrary replied:
Please check lines 180, 181, 195 and 205 of the final file with all patches applied.

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
++
++  ComPtr<IAsyncAction> async_action_media_capture;
++  Event event_wait_for_media_capture_async_action;
++  Event event_wait_for_media_capture_async_action_complition;
```
Declaring variables at the top of the functions is arguably very C-style, which would be just nit-picking if those were not smart pointers which have a critical effect when going out of scope. Here one has to scan the entire function to find where those variables start to be used and if/when they are reused. Can we convert that to more modern practices please and have variables declared when they're effectively needed, with the minimum needed scope?

LoadLibrary replied:
Declaring the variables at the top of the scopes is a consequence of relying on the recommended "Jump on Fail" error handling pattern for COM published by Microsoft (https://docs.microsoft.com/en-us/windows/win32/learnwin32/error-handling-in-com#jump-on-fail).

Keep in mind that the code is not using a projection (like WinRT/C++), it is calling the WinRT ABI directly. The reasoning for that is based on the fact that WinRT/C++ requires C++ exceptions and Google doesn't take code with C++ exceptions (https://google.github.io/styleguide/cppguide.html#Exceptions).

WinRT is based on COM and using an approved COM error handling patterns is a good practice. Won't fix.

LoadLibrary replied:
This change has been updated with the "Cascade of ifs" instead of "Jump on Fail". Although "Cascade of ifs" drops the THR macro and its goto, it doesn't drop the requirement of declaring variable at the beginning of the scope.
Declaring variable near to its use would require the use of the "Nested ifs" or "Throw on fail" patterns. "Nested ifs" is not adequate for this code base due its long chain of API calls. "Throw on fail" can't be used because we can't have exceptions in this code (Google guideline).

fiban-havok replied:
Should be fine now - can be improved IMHO, but consider the following as nitpicking and optional.

Technically "cascade of ifs" does not require declaring all variables at the beginning - you can declare them just above the `if(SUCCEEDED)` block where they are used, which isn't perfect but is better IMHO. It can be argued that having symmetrical blocks for initializations and cleanups is less error-prone ofc.

More importantly, this is (luckily) C++ code, and predominantly (or exclusively - haven't checked all the cases) uses `ComPtr`, so we can arguably do better than the MS docs examples (which are C and use raw pointers needing manual cleanup). I'd say this pattern is better:
```cpp
HRESULT hr;

ComPtr<A> a;
hr = DoSomething(&a);
if (FAILED(hr)) {
  return hr;
}

// Here we do not propagate the error code so can be shorter
ComPtr<B> b;
if (FAILED(DoSomethingElse(&b))) {
  return E_FAIL;
}

if (SomeCondition()) {
  // Variables do not need to be declared at root scope.
  ComPtr<C> c;
  hr = DoSomethingElseEntirely(&c);
  if (FAILED(hr)) {
    return hr;
  }
}
```
Seems the most C++-idiomatic way to me. Thoughts?

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
++  // We'll be dispatching and waiting for code from the UI thread.
++  // Let's make sure this is not the UI thread.
++  THR(main_view_dispatcher->get_HasThreadAccess(&has_thread_access));
++  THR(has_thread_access ? E_FAIL : S_OK);
```
This should probably return `E_WRONGTHREAD` rather than the generic `E_FAIL`.

LoadLibrary replied:
There is no E_WRONGTHREAD in the Windows SDK.
E_FAIL has been replaced by E_RPC_WRONG_THREAD.

jehumb-havok replied:
Right, I was writing from memory, that's the one I was thinking of, thanks.

LoadLibrary replied:
E_RPC_WRONG_THREAD has been adopted instead of E_FAIL.

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
++  // The media capture device has to be initialized in the UI thread.
++  // The following event is used for letting this thread know that
++  // async_action_media_capture has been created.
++  event_wait_for_media_capture_async_action =
```
Use `rtc::Event`.

LoadLibrary replied:
This code will never be cross platform.
What are the benefits of using ::rtc::Event instead of ::Microsoft::WRL::Wrappers::Event for this code?

jehumb-havok replied:
-2 lines of code, consistency of codebase, and familiarity for non-Windows WebRTC devs. I personally have no idea what all those options in `CreateEventEx()` are; to review I need to pop up the documentation.

LoadLibrary replied:
Won't fix. This is a platform specific code and CreateEventEx is the platform API for creating events.

jehumb-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -0,0 +1,695 @@
++            StreamingCaptureMode::StreamingCaptureMode_AudioAndVideo));
++        THR(init_settings.As<IMediaCaptureInitializationSettings5>(
++            &init_settings5));
++        THR(init_settings5->put_MemoryPreference(
```
I assume this will go away with the no-copy optimization, since we want the frame to stay in VRAM until it's encoded, no?

LoadLibrary replied:
This code was designed to be a turnkey solution for WebRTC's built-in video capture module and WebRTC's built-in codecs. WebRTC built-in codecs run in CPU. The implementation that will supersede this one will follow Anaheim's approach (whatever it is).

Keep in mind that not all cameras have direct access to NVRAM. For example, USB Webcams transfer the captured images to RAM instead of VRM. Whatever code I put in place will have to address both use cases.

LoadLibrary reviewed (commented):
All provided feedback has been addressed.

LoadLibrary committed:
Incorporating feedback

Using RPC_E_WRONG_THREAD instead of E_FAIL, using log macro from ::rtc
for reporting function name and fixing comment about stride workaround.

LoadLibrary committed:
Replacing error handling pattern from "Jump On Fail" to "Cascading ifs"

This change deprecates the use of the THR macro (and its goto Cleanup) in favor of the "Cascading ifs" pattern
(https://docs.microsoft.com/en-us/windows/win32/LearnWin32/error-handling-in-com#cascading-ifs).

This change also removes the use of ReleaseInterface in favor of passing refs to ComPtrs.

fiban-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -1,19 +1,34 @@
+
+Removing use of the ReleaseInterface function
+
+help_functions_winrt.cc without the sinful goto
+
+video_capture_winrt.cc deprived of all goto sins
+
+device_info_winrt, repent for the goto sins
+
+No more goto or THR sins
```
nit: I'd stick to "Removed gotos from a, b, and c", this comes off as a bit snarky (not important if we don't plan to contribute this patch)

LoadLibrary replied:
Don't worry. :) Patch files are sanitized before contributing back. The process of contributing back involves applying the patch files to the WebRTC local repo, running the tidy apps and clean up log messages while rebasing.

LoadLibrary committed:
Fixing libyuv NV12-to-I420 conversion with strides and gap

fiban-havok commented on patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch:
```diff
@@ -20,13 +20,93 @@ device_info_winrt, repent for the goto sins
+ 
+@@ -156,7 +151,7 @@ int32_t VideoCaptureImpl::IncomingFrame(uint8_t* videoFrame,
+ 
+   // TODO(nisse): Use a pool?
```
nit: by all means use a pool! The corresponding WebRTC-UWP code was a noticeable perf issue on HL. See https://github.com/webrtc-uwp/webrtc-apis/commit/ced46e4c94b227514c078e7004ee776df8c350ea for a fix.

Not relevant to this PR though so can be done separately, just a reminder.

LoadLibrary replied:
Feel free to file an issue.

fiban-havok replied:
Reported as #32.

fiban-havok commented on patches_for_WebRTC_org/m80/0007.3-Fixing-NV12-to-I420-conversion-with-strides-and-gap.patch:
```diff
@@ -0,0 +1,76 @@
+ LIBYUV_API
+ int ConvertToI420(const uint8_t* sample,
+                   size_t sample_size,
++                  int src_stride_y,
++                  const uint8_t* src_uv,
++                  int src_stride_uv,
```
See Jerome's comment, we should use `NV12ToI420()` and `NV12ToI420()` if we know these are the only formats that we are going to get. If we need to support other formats, we can use the explicit conversion functions for the ones that may have separate Y and UV planes (dunno if there are others apart from NV12 and NV21) and the default `ConvertToI420` for all the others.

This specific change has issues that make it viable just as a short-term workaround (breaks the API of libyuv; adds parameters that are useful only to specific formats to a generic function; maybe implementation is incomplete as Jerome pointed out). Also note that `ConvertToI420` already takes both `src_width` and `crop_x` that can be used as stride and width in general.

LoadLibrary replied:
We shouldn't use `NV12ToI420()` or `NV12ToI420()` directly. There is no guarantee that NV12 is the only format that will be ingested. Besides, these methods do not rotate or crop the image when required like `ConvertToI420` does. 
It's right that this change breaks API and that this is a short-term workaround. This API is already broken because it assumes that the UV plane is adjacent to the Y plane and that the source stride is its width.
I have no plans to contribute back this change. I'd like to stop using libyuv in favor of MediaFoundation as soon as possible.

fiban-havok replied:
Ofc I meant `NV12ToI420()` or <code>NV**21**ToI420()</code> 😉 

> these methods do not rotate or crop the image

There is a `NV12ToI420Rotate` that does take crop and rotation as you can see in this same file (it will work trivially for NV21 too).

> This API is already broken because it assumes that the UV plane is adjacent to the Y plane 

By "break" I meant "break compatibility", not functionality. Also I wouldn't call `ConvertToI420` "broken" - AFAICS  is a generic tool and (by design I assume) doesn't support all possible cases for all possible formats (it would need an unwieldy number of parameters to do that). AFAICS the function works correctly for all supported input.

> and that the source stride is its width.

It does handle images with a larger stride, you need to pass stride as `src_width` and width as `crop_x` as said above.

> I have no plans to contribute back this change. I'd like to stop using libyuv in favor of MediaFoundation as soon as possible.

Fair enough. It seems to me that NV12 and NV21 are the only formats in `ConvertToI420` that need this special handling. I'd say that using `NV12ToI420Rotate` for those and `ConvertToI420` for all the others is easy enough and avoids this kind of workaround so it's preferable. If there are strong reasons to still go for this, you should document that this change is a temporary workaround, that it breaks compatibility with mainline libyuv, and that those parameters are only used in conversion from NV12/NV21.
